bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java

```diff
@@ -0,0 +1,206 @@
+            return component.getName();
+        }
+
+        String getFieldName() {
+            if (component.isAnnotationPresent(BsonId.class)) {
+                return "_id";
+            } else if (component.isAnnotationPresent(BsonProperty.class)) {
+                return component.getAnnotation(BsonProperty.class).value();
+            }
+            return getComponentName();
+        }
+
+        Object getValue(final Record record) throws InvocationTargetException, IllegalAccessException {
+            return component.getAccessor().invoke(record);
+        }
+    }
+
+    RecordCodec(final Class<T> clazz, final CodecRegistry codecRegistry) {
+        this.clazz = notNull("class", clazz);
+        canonicalConstructor = notNull("canonicalConstructor", getCanonicalConstructor(clazz));
+        componentModels = getComponentModels(clazz, codecRegistry);
+        fieldNameToComponentModel = componentModels.stream()
+                .collect(Collectors.toMap(ComponentModel::getFieldName, Function.identity()));
+        componentModelForId = getComponentModelForId(clazz, componentModels);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T decode(final BsonReader reader, final DecoderContext decoderContext) {
+        reader.readStartDocument();
+
+        Object[] constructorArguments = new Object[componentModels.size()];
+        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+            var fieldName = reader.readName();
+            var componentModel = fieldNameToComponentModel.get(fieldName);
+            constructorArguments[componentModel.index] = decoderContext.decodeWithChildContext(componentModel.codec, reader);
+        }
+        reader.readEndDocument();
+
+        try {
+            return (T) canonicalConstructor.newInstance(constructorArguments);
+        } catch (ReflectiveOperationException e) {
+            throw new CodecConfigurationException(format("Unable to invoke canonical constructor of record class %s", clazz.getName()), e);
+        }
+    }
+
+    @Override
+    public void encode(final BsonWriter writer, final T record, final EncoderContext encoderContext) {
+        writer.writeStartDocument();
+        if (componentModelForId != null) {
+            writeComponent(writer, record, componentModelForId);
+        }
+        for (var componentModel : componentModels) {
+            if (componentModel == componentModelForId) {
+                continue;
+            }
+            writeComponent(writer, record, componentModel);
+        }
+        writer.writeEndDocument();
+
+    }
+
+    @Override
+    public Class<T> getEncoderClass() {
+        return clazz;
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private void writeComponent(final BsonWriter writer, final T record, final ComponentModel componentModel) {
+        try {
+            Object componentValue = componentModel.getValue(record);
+            if (componentValue != null) {
```

Comment:
Skips nulls entirely

Reply:
This is different to POJOs that check if values should be serialized or not via `PropertySerialization`

Reply:
Ah, I see.  Seems like the default behavior for POJOs is the same as this, as there is only one implementation of `PropertySerialization` in the driver (`PropertyModelSerializationImpl`), and it has the same behavior:

```
    public boolean shouldSerialize(final T value) {
        return value != null;
    }
```

If we end up adding convention support for records, then something like `PropertySerialization` would likely be included.